text/0374-jsii-ts-version.md

```diff
@@ -0,0 +1,324 @@
+  follow a different versioning scheme.
+
+  - Move other packages to a new mono-repository (or to several new individual
+    package repositories), e.g: `github.com/aws/jsii-toolchain`.
+
+  - Make the `github.com/aws/jsii` repository by a single-package repository.
+
+* Update the *release* automation to use the [TypeScript] compiler version's
+  `major.minor` level, and only update the `patch` level on new releases.
+
+* Upgrade the [TypeScript] compiler internally used by `jsii` to the current
+  `latest` release of the `typescript` package.
+
+* Make the `jsii` compiler transparently prepare down-leveled declarations files
+  to ensure backwards-compatibility of compiler outputs with previous releases
+  of `jsii`. This can be done using the [`downlevel-dts`][downlevel-dts] package
+  to produce declarations files compatible with [TypeScript] `3.9`, and adding
+  a `typesVersions` key to the `package.json` file of packages.
+
+  - See proof-of-concept: [`aws/jsii#3501`][aws/jsii-3501]
+
+* Release the initial `jsii` release on the `4.7` line.
+
+* Formally announce that the `v1` release line of `jsii` will move into the
+  *Maintenance* tier of the [AWS SDKs and Tools maintenance policy][aws-policy]
+  in 6 months, and explain the policy that is applicable to newer releases.
+
+* **6 months later:** Formally announce that the `v1` release line of `jsii` is
+  entering the *Maintenance* tier of the [AWS SDKs and Tools maintenance
+  policy][aws-policy], and will continue to receive critical bug and security
+  fixes for 12 months before transitioning to *end-of-life*, and repeat the
+  maintenance policy that new releases will benefit from.
+
+* **12 months later:** Formally announce that the `v1` release line of `jsii` is
+  transitioning to *end-of-life* and will no longer receive any updates. Repeat
+  the maintenance policy that is applicable to newer releases.
+
+### Is this a breaking change?
+
+This is not a breaking change in the traditional sense, however this is a
+significant change in versioning schemes used for the `jsii` compiler, and a
+departure from the [AWS SDKs and Tools maintenance policy][aws-policy].
+
+As such, this warrants ample communication with our customer base to avoid they
+are caught by surprise, or have broken expectations.
+
+### What alternative solutions did you consider?
+
+#### Bring-your-own TypeScript version
+
+One of the main rationales for this change is to enable the use of new
+TypeScript releases without forcing the enitre package ecosystem to make the
+change at the same time (due to language-breaking changes).
+
+An option to address this would be to allow customers to "bring their own"
+TypeScript compiler, by making it a `peerDependency`. This however comes with
+a significant challenge: the TypeScript compiler API is not stable between
+TypeScript releases, and it is hence not possible to author a `jsii` compiler
+that can leverage an arbitrary TypeScript version.
```

Comment:
But it is though. We could have different adapters for different TS versions:

```js
// package.json
{
  "devDependencies": {
     "typescript-39": "npm:typescript@~3.9.0",
     "typescript-47": "npm:typescript@~4.7.0"
  }
}
```

```ts
// typescript-39.ts
import type * as ts from 'typescript-39';

class Typescript39Compiler implements ICompiler {
  private readonly ts: typeof import 'typescript-39';

  constructor() {
    this.ts = requireFromUserScope('typescript');

    const program = new this.ts.Program(...);
  }
}
```

Reply:
That's an interesting proposition... I'm going to be investigating that one a bit more...

Instinctively, I feel we might have an issue due to how the `ts.Node` types (and descendants) might not be compatible between versions, and the `jsii.Assembler` currently works by traversing the AST and gathering information... The feasibility of this approach hinges on our ability to design an abstraction that hides those differences effectively while not forcing us to effectively maintain a distinct assembler per compiler version.

Reply:
That'll be a rejected alternative. The TypeScript compiler fights against this idea pretty strongly: writing an adapter layer almost boils down to re-writing a unified front-end to the entire TypeScript compiler surface by hand (it's almost impossible to re-use common signatures, in parts due to how the TypeScript compiler exports a mix of values & pure types).

I'll see to add a paragraph describing the problem here...